backend/hyper/hypersql/queries.go

```diff
@@ -287,6 +328,31 @@ SELECT json_group_array(blob) AS heads
 FROM non_drafts
 WHERE blob NOT IN deps`,
 		},
+		sgen.QueryTemplate{
+			Name: "ChangesGetTrustedHeadsJSON",
+			Kind: sgen.QueryKindSingle,
+			Inputs: []sgen.GoSymbol{
+				{Name: "eid", Type: sgen.TypeText},
+			},
+			Outputs: []sgen.GoSymbol{
+				{Name: "Heads", Type: sgen.TypeBytes},
+			},
+			SQL: `WITH RECURSIVE
```

Comment:
I've spent more then 30 minutes trying to understand this query, and I'm still confused by the nested CTE, the recursiveness of it, and the `LIMIT 100000`.

I just realized something that could dramatically simplify all of this.

The process to load any entity is roughly 3 steps:

1. Find the heads.
2. Resolve the heads: i.e. walk recursively all the dependencies from heads up to the initial change.
3. Load the changes from the resolved change set and apply them to the entity.

So, what I realized is that we can pass redundant changes to the second step, no need to pass only "pure" heads. E.g. if we have changes `A <- B <- C`, we can pass `[B,C]` to the resolve step instead of just `[C]`. It will walk up all the dependencies recursively, and deduplicate everything, so in the end we'll get our `[A,B,C]` correctly.

Based on that realization, seems like the easiest way to load a trusted entity is to load all the trusted changes, and pass them to the resolve step right away. If any untrusted change is somewhere in the dependency chain, we'll get it. If not, we'll get the same list of changes that we passed for resolution, so all good.

I suggest we do that, as it seems to be much easier to reason about. Maybe we need to rename heads to something else in the code, to reflect that the resolve step can work with redundant changes and what it needs is just a "starting point" for recursive traversal.

I just tried this approach, and all the tests pass correctly.

Here's the query I used:

```sql
-- Again, these are not real heads here, just the starting point for the resolve step.
-- The real heads can be obtained easily those using the Entity with already applied changes.
SELECT json_group_array(hd_changes.blob) AS heads
FROM hd_changes
JOIN trusted_accounts ON trusted_accounts.id = hd_changes.author
WHERE hd_changes.entity = :entity
```

Let me know if you also think it's easier this way, and I can make the changes myself.

Reply:
The limit is because in the documentation it says that you should include a limit (in case there is a circular dependency problem I guess and the query runs forever), and I just thought we will never have an entity with such number of ancestors.
The nested CTE works as follows:
Given an initial query that gets the parent of a trusted change: `SELECT DISTINCT hd_change_deps.parent FROM hd_change_deps JOIN trusted_changes ON trusted_changes.blob = hd_change_deps.child` (If we only had that query, then only one parent change to a trusted head will be found) we introduce the nesting by referencing back `trusted_parent` function for every new parent found `UNION SELECT hd_change_deps.parent FROM hd_change_deps, trusted_parent WHERE child = blob LIMIT 100000)`
As for the changes you mention, I don't see the performance benefit, since you still need to walk all the ancestors. With the current algorithm and the example you gave in which we have `A <- B <- C` `C` is going to be the initial one and then `A` and `B` are going to be discovered by recursiveness. But if you thing it makes more sense to change it to be more coherent with more parts of the code, feel free to change it @burdiyan

Reply:
Oh, now I see it, the magic actually happens inside `loadFromHeads`. I should have looked there more carefully instead of treating it like a black box. This is the sensible approach we should have taken from the beginning, but hey! now I think I understand heads much better! and SQL recursives :joy: